llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp

```diff
@@ -0,0 +1,462 @@
+  }
+}
+
+/// Propagates aspects from leaves up to the top of call graph.
+/// NB! Call graph corresponds to call graph of SYCL code which
+/// can't contain recursive calls. So there can't be loops in
+/// a call graph. But there can be path's intersections.
+void propagateAspectsThroughCG(Function *F, CallGraphTy &CG,
+                               FunctionToAspectsMapTy &AspectsMap,
+                               SmallPtrSet<Function *, 16> &Visited) {
+  auto It = CG.find(F);
+  if (It == CG.end())
+    return;
+
+  AspectWithFunctionLinkSetTy LocalAspects;
+  for (auto Edge : It->second) {
+    Function *Callee = Edge.first;
+    if (Visited.insert(Callee).second)
+      propagateAspectsThroughCG(Callee, CG, AspectsMap, Visited);
+
+    auto &CalleeAspects = AspectsMap[Callee];
+    for (auto AspectIt : CalleeAspects) {
+      LocalAspects.insert(
+          AspectWithFunctionLinkTy{AspectIt.Aspect, Callee, Edge.second});
+    }
+  }
+
+  AspectsMap[F].insert(LocalAspects.begin(), LocalAspects.end());
+}
+
+/// Processes function's instructions. It analyzes aspect usages with debug
+/// information and builds a call graph.
+void processFunctionInstructions(Function &F,
+                                 FunctionToAspectsMapTy &FunctionToAspects,
+                                 TypeToAspectsMapTy &TypesWithAspects,
+                                 CallGraphTy &CG) {
+  auto &AspectsWithFunctionLinkSet = FunctionToAspects[&F];
+  for (Instruction &I : instructions(F)) {
+    AspectsSetTy Aspects = getAspectsUsedByInstruction(I, TypesWithAspects);
+    for (int Aspect : Aspects) {
+      FunctionToAspects[&F].insert(
+          AspectWithFunctionLinkTy{Aspect, nullptr, &I.getDebugLoc()});
+    }
+
+    if (auto *DBI = dyn_cast<DbgVariableIntrinsic>(&I)) {
+      // Handle a group of llvm.dbg.{addr,declare,value} intrinsics.
+      // Example:
+      // %tmp = alloca %Struct
+      // call void @llvm.dbg.declare(metadata %Struct* %tmp, metadata !1,
+      //                             metadata !DIExpression), !dbg !2
+      //
+      // Here we extract the first intrinsic argument, then extract
+      // %Struct type, then analyze known aspects of %Struct and update
+      // corresponding aspect's records with debug information !dbg !2.
+      Type *T = DBI->getVariableLocationOp(0)->getType();
+      Aspects = getAspectsFromType(T, TypesWithAspects);
+      for (int Aspect : Aspects) {
+        AspectWithFunctionLinkTy AspectWithFunctionLink{Aspect};
+        auto It =
+            std::find(AspectsWithFunctionLinkSet.begin(),
+                      AspectsWithFunctionLinkSet.end(), AspectWithFunctionLink);
+        if (It == AspectsWithFunctionLinkSet.end())
+          continue;
+
+        AspectWithFunctionLink = *It;
+        AspectsWithFunctionLinkSet.erase(AspectWithFunctionLink);
+        AspectWithFunctionLink.DL = &I.getDebugLoc();
+        AspectsWithFunctionLinkSet.insert(AspectWithFunctionLink);
```

Comment:
Why do we need `erase` + `insert`, why can't we just update `.DL` field?

Reply:
`AspectWithFunctionLinkSet` is a SmallSet. No set-like container allows to modify its element since order might be broken. However, I can change SmallSet to some map container.